### PR TITLE
Multi imports in one line

### LINKED ISSUE #1742: [0.13.7] import import sbt._, Keys._ breaks build
originally reported by @tkawachi as https://groups.google.com/d/msg/sbt-dev/YQvtYQhOgVk/cCXc1FMiHVkJ
## steps

sbt 0.13.7

``` scala
import sbt._, Keys._

scalaVersion := "2.11.4"
```
## problem

Build breaks

### PR DESCRIPTION
**Pull request by ajozwik**
Fixes #1742

Correction for imports like: `import sbt._, Keys._,java.util._`

### CODE REVIEW

**Comment by typesafe-tools** [comment]
Can one of the admins verify this patch?

**Comment by jsuereth** [comment]
Hmm, tests are failing: https://travis-ci.org/sbt/sbt/jobs/41653158#L5420

**Comment by ajozwik** [comment]
Now the problem is with dependencies, see https://travis-ci.org/sbt/sbt/jobs/41690279#L4715

**Comment by ajozwik** [comment]
I corrected the parser (for xml) - I added the class ScalaParser.scala for parsing sbt file. If xml is in comment - it should not be touched.
This parser does not handle all cases - it is used as helper to handle xml parts (original scala parser do not support it)

File: main/src/main/scala/sbt/internals/parser/SbtParser.scala [hunk 1]
```diff
@@ -145,7 +143,46 @@ private[sbt] case class SbtParser(file: File, lines: Seq[String]) extends Parsed
...
+    (importsLineRange, stmtTreeLineRange.map { case (stmt, _, lr) => (stmt, lr) }, stmtTreeLineRange.map { case (stmt, tree, _) => (stmt, tree) }, modifiedContent)
+  }
+
+  /**
+   * import sbt._, Keys._,java.util._ should return ("import sbt._, Keys._,java.util._",0)
+   * @param modifiedContent - modifiedContent
+   * @param imports - trees
+   * @return imports per line
+   */
+  private def importsToLineRanges(modifiedContent: String, imports: Seq[Tree]): Seq[(String, Int)] = {
```
Comment (jsuereth) [inline main/src/main/scala/sbt/internals/parser/SbtParser.scala:155]: So, interestingly, this is what we almost did to fix the `val x,y = 5` issue..... cc> @eed3si9n

File: main/src/main/scala/sbt/internals/parser/SbtParser.scala [hunk 2]
```diff
@@ -267,7 +304,14 @@ private[sbt] object XmlContent {
    * @return sorted by openIndex xml parts
    */
   private def findXmlParts(content: String): Seq[(String, Int, Int)] = {
-    val xmlParts = findModifiedOpeningTags(content, 0, Seq.empty) ++ findNotModifiedOpeningTags(content, 0, Seq.empty)
+    val characterTypeCharSeq = util.Try(ScalaParser(content).characterTypeCharSeq) match {
+      case util.Success(p) =>
+        p
+      case util.Failure(th) =>
+        println(th.getMessage)
```
Comment (jsuereth) [inline main/src/main/scala/sbt/internals/parser/SbtParser.scala]: we won't want the println in the final.  Maybe we can pass a logger here?

File: main/src/main/scala/sbt/internals/parser/ScalaParser.scala [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+  val COMMENTED = Value
+
+  // scala variable like `throw`
+  val BACKTICK = Value
+
+  // any of above
+  val NORMAL = Value
+}
+
+object ScalaParser {
```
Comment (jsuereth) [inline main/src/main/scala/sbt/internals/parser/ScalaParser.scala]: This appears to be more of a Lexer/Tokenizer (with `CharacterType` being the tokens).

File: main/src/test/scala/sbt/internals/parser/XmlFormatSpec.scala [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+package internals.parser
+
+import java.io.File
+
+import org.specs2.matcher.MatchResult
+import org.specs2.mutable.Specification
+
+import scala.io.Source
+
+class XmlFormatSpec extends AbstractInputFolderSpecification("xml-format", "Do not change content when xml is in string ", NameFilter.fnToNameFilter(_.endsWith("sbt.txt"))) {
```
Comment (jsuereth) [inline main/src/test/scala/sbt/internals/parser/XmlFormatSpec.scala]: very much appreciate the new tests.

File: main/src/main/scala/sbt/internals/parser/SbtParser.scala (hunk 2 above)
Comment (ajozwik) [inline main/src/main/scala/sbt/internals/parser/SbtParser.scala]: Replace println with

``` scala
ConsoleLogger(System.out).error(th.getMessage)
```

File: main/src/main/scala/sbt/internals/parser/ScalaParser.scala (hunk 3 above)
Comment (ajozwik) [inline main/src/main/scala/sbt/internals/parser/ScalaParser.scala]: CharacterType => TokenType ?

ScalaParser => ScalaTokenizer ?

File: main/src/test/scala/sbt/internals/parser/XmlFormatSpec.scala (hunk 4 above)
Comment (ajozwik) [inline main/src/test/scala/sbt/internals/parser/XmlFormatSpec.scala]: Spec extends abstract class. Abstract class contains method to read from `folder` files filtered by `fileFilter` and run method `checkFile` for each file. XmlFormatSpec contains only implementation - checkFile method.
The input files are in `folder` directory - here in `xml-format` 
`main/src/test/resources/xml-format`

**Comment by eed3si9n** [comment]
Do you want to roll back to ajozwik/sbt@ce75c5a on this branch or make a new PR?

**Comment by ajozwik** [comment]
@eed3si9n If you can merge ajozwik/sbt@ce75c5a it will be nice. I red documentation about remove commit from github, but I think it is risky.
What is a easy way to roll back to ce75c5a?

**Comment by eed3si9n** [comment]
You're working on 0.13 branch, but it's a good practice to always work on a topic branch and use `0.13` only for tracking upstream (sbt/sbt).

```
$ git checkout -b backup/0.13
$ git log
$ git checkout -b fix/1742
$ git log
$ git reset --hard HEAD^^^
```

The above will create new branches called `backup/0.13` and `fix/1742` out of your current `0.13`, and remove 3 commits off the HEAD. You can close this PR and re-send it from that one.

**Comment by ajozwik** [comment]
Check now, I think I remove commits. I have local copy with xml correction. I will create new PR for xml - I think the correction in scala takes some time.

**Pull request merged**